### PR TITLE
fix demo for dependencyInterpretation

### PR DESCRIPTION
**Pull request by chaosi-zju**
**What type of PR is this?**

/kind documentation
/kind cleanup

**What this PR does / why we need it**:

This demo is written as a reference for users, but it is not actually available, which will make users think that our feature has a bug.

here are two errors:

1、field name fault:
 * `desiredObj.spec.serviceAccountName` -> `desiredObj.spec.template.spec.serviceAccountName`
 * `desiredObj.namespace` -> `desiredObj.metadata.namespace`

2、array index error:
* `dependencies[0] = dependency` -> `dependencies[1] = dependency` (should not use index 0)
otherwise you will get the error:

```console
E0529 15:09:13.004066 1245422 configurable.go:177] Failed to get DependentObjectReferences from object: apps/v1, Kind=Deployment default/nginx, error: encode lua result to json failed: json: error calling MarshalJSON for type json.jsonValue: cannot encode sparse array
Execute InterpretDependency error: encode lua result to json failed: json: error calling MarshalJSON for type json.jsonValue: cannot encode sparse array
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by chaosi-zju** [comment]
As for error 2 `array index error` mentioned above, below is test method.

## Before bug fixed

1. deployment.yaml
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  namespace: default
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      terminationGracePeriodSeconds: 0
      serviceAccountName: admin
      containers:
        - image: nginx
          name: nginx
          resources:
            limits:
              cpu: 10m
              memory: 10Mi
```
2. interpreter.yaml
```yaml
apiVersion: config.karmada.io/v1alpha1
kind: ResourceInterpreterCustomization
metadata:
  name: deployment-dependency-hpa
spec:
  target:
    apiVersion: apps/v1
    kind: Deployment
  customizations:
    dependencyInterpretation:
      luaScript: >
        function GetDependencies(desiredObj)
          dependencies = {}
          serviceAccountName = desiredObj.spec.template.spec.serviceAccountName
          if serviceAccountName ~= nil and serviceAccountName ~= "default" then
            dependency = {}
            dependency.apiVersion = "v1"
            dependency.kind = "ServiceAccount"
            dependency.name = serviceAccountName
            dependency.namespace = desiredObj.metadata.namespace
            dependencies[0] = {}
            dependencies[0] = dependency
          end
          return dependencies
        end
```
3. run the command:
```shell
$ karmadactl --karmada-context karmada-apiserver interpret -f interpreter.yaml --operation interpretDependency --observed-file deployment.yaml
E0529 15:09:13.004066 1245422 configurable.go:177] Failed to get DependentObjectReferences from object: apps/v1, Kind=Deployment default/nginx, error: encode lua result to json failed: json: error calling MarshalJSON for type json.jsonValue: cannot encode sparse array
Execute InterpretDependency error: encode lua result to json failed: json: error calling MarshalJSON for type json.jsonValue: cannot encode sparse array
```

## After bug fixed
1. deployment.yaml
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  namespace: default
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      terminationGracePeriodSeconds: 0
      serviceAccountName: admin
      containers:
        - image: nginx
          name: nginx
          resources:
            limits:
              cpu: 10m
              memory: 10Mi
```
2. interpreter.yaml
```yaml
apiVersion: config.karmada.io/v1alpha1
kind: ResourceInterpreterCustomization
metadata:
  name: deployment-dependency-hpa
spec:
  target:
    apiVersion: apps/v1
    kind: Deployment
  customizations:
    dependencyInterpretation:
      luaScript: >
        function GetDependencies(desiredObj)
          dependencies = {}
          serviceAccountName = desiredObj.spec.template.spec.serviceAccountName
          if serviceAccountName ~= nil and serviceAccountName ~= "default" then
            dependency = {}
            dependency.apiVersion = "v1"
            dependency.kind = "ServiceAccount"
            dependency.name = serviceAccountName
            dependency.namespace = desiredObj.metadata.namespace
            dependencies[1] = dependency
          end
          return dependencies
        end
```
3. run the command:
```shell
$ karmadactl --karmada-context karmada-apiserver interpret -f interpreter.yaml --operation interpretDependency --observed-file deployment.yaml
---
# [1/1] dependencies:
- apiVersion: v1
  kind: ServiceAccount
  name: admin
  namespace: default
```

**Comment by XiShanYongYe-Chang** [comment]
Thanks a lot 
/assign

File: pkg/util/interpreter/rule.go [hunk 1]
```diff
@@ -367,14 +367,14 @@ func (d *dependencyInterpretationRule) Document() string {
 The script should implement a function as follows:
 function GetDependencies(desiredObj)
   dependencies = {}
-  if desiredObj.spec.serviceAccountName ~= nil and desiredObj.spec.serviceAccountName ~= "default" then
```
Comment (chaunceyjiang) [inline pkg/util/interpreter/rule.go:370]: I believe this is not a bug. It's more like an update to the documentation.

Comment (XiShanYongYe-Chang) [reply pkg/util/interpreter/rule.go:370]: +1

There are other places throughout the codebase that need to be changed as well.

**Review by chaunceyjiang — COMMENTED** [review]
Thanks

/lgtm

**Comment by chaosi-zju** [comment]
@XiShanYongYe-Chang any more comments?

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks
/approve

**Pull request merged**